pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go

```diff
@@ -0,0 +1,527 @@
+	for nodeName, pods := range nodeToDaemonPods {
+		// Check if node is ready, ignore not-ready node
+		ready, err := NodeReadyByName(c.nodeLister, nodeName)
+		if err != nil {
+			return fmt.Errorf("couldn't check node %q ready status, %v", nodeName, err)
+		}
+		if !ready {
+			continue
+		}
+		for _, pod := range pods {
+			if err := SetPodUpgradeCondition(c.kubeclientset, ds, pod); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// autoUpdate identifies the set of old pods to delete within the constraints imposed by the max-unavailable number.
+// Just ignore and do not calculate not-ready nodes.
+func (c *Controller) autoUpdate(ds *appsv1.DaemonSet, nodeToDaemonPods map[string][]*corev1.Pod) error {
+	// Calculate maxUnavailable specified by user, default is 1
+	maxUnavailable, err := c.maxUnavailableCounts(ds, nodeToDaemonPods)
+	if err != nil {
+		return fmt.Errorf("couldn't get maxUnavailable number for daemon set %q: %v", ds.Name, err)
+	}
+
+	var numUnavailable int
+	var allowedReplacementPods []string
+	var candidatePodsToDelete []string
+
+	for nodeName, pods := range nodeToDaemonPods {
+		// Check if node is ready, ignore not-ready node
+		// this is a significant difference from the native daemonset controller
+		ready, err := NodeReadyByName(c.nodeLister, nodeName)
+		if err != nil {
+			return fmt.Errorf("couldn't check node %q ready status, %v", nodeName, err)
+		}
+		if !ready {
+			continue
+		}
+
+		newPod, oldPod, ok := findUpdatedPodsOnNode(ds, pods)
+		if !ok {
+			// Let the manage loop clean up this node, and treat it as an unavailable node
+			klog.V(3).Infof("DaemonSet %s/%s has excess pods on node %s, skipping to allow the core loop to process", ds.Namespace, ds.Name, nodeName)
+			numUnavailable++
+			continue
+		}
+		switch {
+		case oldPod == nil && newPod == nil, oldPod != nil && newPod != nil:
+			// The manage loop will handle creating or deleting the appropriate pod, consider this unavailable
+			numUnavailable++
+		case newPod != nil:
+			// This pod is up-to-date, check its availability
+			if !k8sutil.IsPodAvailable(newPod, ds.Spec.MinReadySeconds, metav1.Time{Time: time.Now()}) {
+				// An unavailable new pod is counted against maxUnavailable
+				numUnavailable++
+			}
+		default:
+			// This pod is old, it is an update candidate
+			switch {
+			case !k8sutil.IsPodAvailable(oldPod, ds.Spec.MinReadySeconds, metav1.Time{Time: time.Now()}):
+				// The old pod isn't available, so it needs to be replaced
+				klog.V(5).Infof("DaemonSet %s/%s pod %s on node %s is out of date and not available, allowing replacement", ds.Namespace, ds.Name, oldPod.Name, nodeName)
+				// Record the replacement
+				if allowedReplacementPods == nil {
+					allowedReplacementPods = make([]string, 0, len(nodeToDaemonPods))
+				}
+				allowedReplacementPods = append(allowedReplacementPods, oldPod.Name)
+			case numUnavailable >= maxUnavailable:
+				// No point considering any other candidates
+				continue
```

Comment:
Can you return directly and wait for next sync?

Reply:
Referring to the K8s `daemonset controller`, old unavailable pods do not count in the maxUnavailable number, so do not return directly here may be better.

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/update.go#L69